roles/download/tasks/download_container.yml

```diff
@@ -1,40 +1,125 @@
 ---
-- name: container_download | Make download decision if pull is required by tag or sha256
-  include_tasks: set_docker_image_facts.yml
+- name: download_container | Set a few facts
+  import_tasks: set_container_facts.yml
+  run_once: "{{ download_run_once }}"
+  tags:
+    - facts
+
+- name: download_container | Determine if image is in cache
+  stat:
+    path: "{{ image_path_cached }}"
+  delegate_to: localhost
+  delegate_facts: no
+  register: cache_image
+  changed_when: false
+  become: false
   when:
-    - download.enabled
-    - download.container
+    - download_force_cache
+
+- name: download_container | Set fact indicating if image is in cache
+  set_fact:
+    image_is_cached: "{{ cache_image.stat.exists | default(false) }}"
   tags:
     - facts
 
-# FIXME(mattymo): In Ansible 2.4 omitting download delegate is broken. Move back
-#                 to one task in the future.
-- name: container_download | Download containers if pull is required or told to always pull (delegate)
-  command: "{{ docker_bin_dir }}/docker pull {{ pull_args }}"
-  register: pull_task_result
-  until: pull_task_result is succeeded
+- name: download_container | Upload image to node if it is cached
+  synchronize:
+    src: "{{ image_path_cached }}"
+    dest: "{{ image_path_final }}"
+    use_ssh_args: "{{ has_bastion | default(false) }}"
+    mode: push
+  delegate_facts: no
+  register: upload_image
+  failed_when: not upload_image
+  run_once: "{{ download_run_once }}"
+  until: upload_image is succeeded
   retries: 4
   delay: "{{ retry_stagger | random + 3 }}"
-  changed_when: not 'up to date' in pull_task_result.stdout
   when:
-    - download_run_once
-    - download.enabled
-    - download.container
-    - any_pull_required | default(download_always_pull)
-  delegate_to: "{{ download_delegate }}"
-  delegate_facts: yes
-  run_once: yes
+    - download_force_cache
+    - image_is_cached
+    - not download_localhost
+    - ansible_os_family not in ["CoreOS", "Container Linux by CoreOS"]
+  tags:
+    - upload
+    - upgrade
+
+# NOTE: If download_delegate==localhost, then the account running ansible must be in the group 'docker' on the ansible host
+- name: download_container | Load image into docker
+  shell: "{{ docker_bin_dir }}/docker load < {{ image_path_cached if download_localhost else image_path_final }}"
+  delegate_to: "{{ download_delegate if download_run_once or inventory_hostname }}"
+  run_once: "{{ download_run_once }}"
+  register: container_load_status
+  failed_when: container_load_status | failed
+  become: "{{ not download_localhost }}"
```

Comment:
why the negation ? what am i missing here ?

Reply:
When running locally, you shouldn't need root to access docker. The user should be a member of the docker group. This prevents having to configure passwordless sudo or having to use ask-become-password. On remote nodes though, passwordless root is not an issue, as lots of stuff is already done as root. In summary: Use become if not running on localhost. Good remark, because I noticed it should be this:  become: "{{ not (download_run_once and download_localhost) }}" . Done.

Reply:
have you pushed a change ?
i don't see ` become: "{{ not (download_run_once and download_localhost) }}"`

Reply:
I've not pushed yet. I'll push when I've figured out & fixed why the coreOS build is failing, because this PR isn't going anywhere if that is not fixed. But I badly need some sleep now, so I'll hopefully find some time this week to spend on it.

Reply:
Pushed it. CI is now failing on "Wait until SSH is available". Bug in the pipeline it seems. I have to push another change anyway, so it'll re-run the tests then.